site_scons/site_tools/compiler.py

```diff
@@ -0,0 +1,131 @@
+"""swig_scanner
+
+Special tool wheich defines compiling/linking options.
+
+This tool is supposed to handle all combinations of OS and compiers.
+When new toolset is added this tool may need to be extended.
+
+For some details:
+
+https://jira.lsstcorp.org/browse/DM-3662
+"""
+
+import os
+import re
+import subprocess
+import sys
+
+import SCons
+import state   # this should not be here
+
+def _compiler(env):
+    """
+    Guess compiler type and version. Returns pair of strings:
+    - toolchain name, one of 'gcc', 'clang'
+    - compiler + version, e.g. 'gcc44', 'gcc48', 'clang35'
+    """
+
+    # for qserv we only set or use CXX, ignore CC
+    comp = os.path.basename(env['CXX'])
+
+    # borrowed from sconsUtils, list of version strings and their matching compiler
+    version_strings = (
+        (r"g(?:\+\+|cc) +\(.+\) +([0-9.a-zA-Z]+)", "gcc"),
+        (r"LLVM +version +([0-9.a-zA-Z]+) ", "clang"), # clang on Mac
+        (r"clang +version +([^ ]+) ", "clang"), # clang on linux
+        (r"\(ICC\) +([0-9.a-zA-Z]+) ", "icc"),
+    )
+
+    # run $comp --version
+    try:
+        output = subprocess.check_output([env['CXX'], '--version'], stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError as exc:
+        # process ran but returned error code
+        raise SCons.Errors.StopError('Error while runnig CXX (%s --version): %s' % (env['CXX'], str(exc)))
+    except Exception as exc:
+        # process did not run
+        raise SCons.Errors.StopError('Failed to execute CXX (%s): %s' % (env['CXX'], str(exc)))
+
+    # try to match output to expected, see that compiler names match
+    toolchain = ''
+    version = ''
+    for re_str, compiler_name in version_strings:
+        match = re.search(re_str, output)
+        if match:
+            toolchain = compiler_name
+            version = match.group(1)
+            break
+
+    # take two first pieces of the version
+    version = re.split('[-.]+', version) + ['']
+    version = ''.join(version[:2])
+
+    return toolchain, toolchain + version
+
+
+def generate(env):
+
+    # current platform
+    platform = sys.platform
+
+    # compiler
+    toolchain, comp_version = _compiler(env)
+
+    state.log.info("Building on %s with %s" % (platform, toolchain))
```

Comment:
We do need better logging there, it's in my long list of improvements (unless we go cmake route).